### PR TITLE
Refactor gas metering.

### PR DESCRIPTION
**Pull request by pepyakin**

### CODE REVIEW

File: src/gas.rs [hunk 1]
```diff
@@ -13,11 +13,71 @@ pub fn update_call_index(opcodes: &mut elements::Opcodes, inserted_index: u32) {
...
+			start_pos: cursor,
+			cost: 1,
+		});
+		self.stack.push(block_idx);
+	}
+
+	/// Seal the current block.
+	///
+	/// Sealed blocks have final cost which will not change later.
+	fn seal(&mut self) {
```
Comment (NikVolf) [inline src/gas.rs]: better name would be finalize, no? Sealing is about integrity

Comment (pepyakin) [reply src/gas.rs]: I think `seal` is fine. In fact, the name is borrowed from the similar domain.

But i'm also good with finalize and could change if you insist

File: src/gas.rs [hunk 2]
```diff
@@ -67,62 +127,66 @@ pub fn inject_counter(
...
-			},
 		}
+
+		cursor += 1;
+	}
+
+	// Then insert metering calls.
+	let mut cumulative_offset = 0;
+	for block in counter.blocks {
+		println!("block {:?}, cumulative_offset={}", block, cumulative_offset);
```
Comment (NikVolf) [inline src/gas.rs]: stray `println` ? :)

Comment (pepyakin) [reply src/gas.rs]: fuuuuu

File: src/gas.rs (hunk 1 above)
Comment (NikVolf) [reply src/gas.rs]: what similar domain?

Comment (pepyakin) [reply src/gas.rs]: There is an algorithm described in the paper some paper `Simple and Efficient Construction of Static Single Assignment Form`. TLDR; This alogrithm processes some code block by block. When all predecessors became known for block this block is sealed, meaning no more predecessors could be found.

It seemed to me that there is analogy to what we are doing here.

Anyway, as I said I'm good with either.

**Pull request merged**